README.md
# Jarolift, SOMFY and RCSwitch_MQTT

Controlling Jarolift(TM) TDEF 433MHz radio shutters and SOMFY radio shutters via ESP8266 and CC1101 Transceiver Module in asynchronous mode.
Receive and Send RF 433 signal ( RCSwitch library)
Experimental version.

Use at your own risk. For private/educational use only. (Keeloq algorithm licensed only to TI Microcontrollers)
This project is not affiliated in any way with the vendor of the Jarolift components.
Jarolift is a Trademark of Schöneberger Rolladenfabrik GmbH & Co. KG

This is the MQTT version of Jarolift_RX_TX_FHEM, which was originally developed to operate with the 
FHEM home automation server. You find the original FHEM version on the project's homepage.


### Original Author
Written by Steffen Hille in Nov, 2017

### Project Homepage and Forum

The project home is here: [Project Home](http://www.bastelbudenbuben.de/2017/04/25/protokollanalyse-von-jarolift-tdef-motoren/)
If you need support please use the forum [Project Forum](http://www.bastelbudenbuben.de/forum/)


### License
The main code of Jarolift_MQTT is licensed under the GPLv3. 
The provided libraries have different licenses, look into the respective files for more information.
* ELECHOUSE CC1101 library	MIT license
* DoubleResetDetector  	MIT License
* Keeloq library      	unknown License
* PubSubClient        	MIT License
* simpleDSTadjust     	unknown License
* RCSwitch             unknown License

### Necessary Hardware:

* a NodeMCU board or similar, based on ESP8266 microcontroller
* a CC1101 Transmitter board (868 Mhz to switch between 433.42 and 44.92 Mhz)
* some connection cables

Photos and hints on project home website
 
### Prepare Arduino IDE and libraries

Download an install the Arduino IDE from [Arduino website](//www.arduino.cc/en/Main/Software)
(as time of writing the current version is 1.8.5 seems have issue view latest release)

Locate your arduino sketch directory - this is `$HOME/Arduino/` in linux

Download and / or unpack this project folder to a subfolder of this sketch directory

Locate the user library folder of your arduino sketch directory

Default under linux: `$HOME/Arduino/libraries/`

* copy the folder `DoubleResetDetector` to the arduino library folder
* copy the folder `PubSubClient` to the arduino library folder
* copy the folder `KeeloqLib` to the arduino library folder
* copy the folder `simpleDSTadjust` to the arduino library folder

### when updating from earlier Jarolift_MQTT releases

* in earlier releases, the folder `PubSubClient-2.6.09` was delivered. Make sure you remove `PubSubClient-2.6.09` from the library folder and put the current `PubSubClient` folder there!


### Compile

Open the Arduino IDE and open the project .ino file (`Jarolift_MQTT.ino`)
TODO write a more detailed description

#### Check the board settings
In menu "Tools" set the correct board type. If you have any kind of NodeMCU, the board
``NodeMCU 1.0 (ESP-12E Module`` will probably work.


You may look at the wiki for [pictures of the modules](https://github.com/madmartin/Jarolift_MQTT/wiki/Devices) or at this [comparison of boards](https://frightanic.com/iot/comparison-of-esp8266-nodemcu-development-boards/).

There are two possible pitfalls when choosing the wrong board type or changing between board types:
* the LED on the board may work / not work - necessary for the LED heardbeat and Admin-Mode.
* take a look at the flash size and the SPIFFS setting - a changed SPIFFS setting needs a new upload of the SPIFFS data files and an existing configuration may be not be found on next boot, then it is initialized with default values.

Note on the LED on the board:

The following settings have proofed to work:

##### ESP Core 2.4.1 / NodeMCU v1.0/V2
* Board: "NodeMCU 1.0 (ESP-12E Module)" or "Generic ESP8266 Module"
* Flash Size: "4M (1M SPIFFS)"
* lwIP variant: "v2 Lower Memory"
* CPU Frequency: "80 MHz"
* Upload speed: "115200"

##### ESP Core 2.4.1 / NodeMCU v3 and WeMos mini & its clones
* Board: "WeMos mini D1 R2 & mini"
* Flash Size: "4M (1M SPIFFS)"
* lwIP variant: "v2 Lower Memory"
* CPU Frequency: "80 MHz"
* Upload speed: "115200"

Dont forget to set the correct serial port (Tools->Port) after plugging the USB connector of the NodeMCU into the computer.

### Uploading files to SPIFFS (ESP8266 internal filesystem)

*ESP8266FS* is a tool which integrates into the Arduino IDE. It adds a
menu item to *Tools* menu for uploading the contents of sketch data
directory into ESP8266 flash file system.

-  Download the [SPIFFS tool](https://github.com/esp8266/arduino-esp8266fs-plugin/releases/download/0.3.0/ESP8266FS-0.3.0.zip)
-  In your Arduino sketchbook directory, create `tools` directory if
   it doesn't exist yet
-  Unpack the tool into `tools` directory (the path will look like
   `$HOME/Arduino/tools/ESP8266FS/tool/esp8266fs.jar`)
-  Restart Arduino IDE
-  Open the Jarolift_MQTT sketch
-  Make sure you have selected a board, port, and closed Serial Monitor
-  Select Tools > ESP8266 Sketch Data Upload. This should start
   uploading the files into ESP8266 flash file system. When done, IDE
   status bar will display `SPIFFS Image Uploaded` message.

Upload the data files (html files)
 Tools / ESP8266 Sketch data upload
Upload the program
 Sketch / Upload
 
### Cabling instructions
ESP | CC1101 | Remark
------|------|------
D8 | CSN
D2 | GDO0 | (perhaps 1Kohm to GND)
D1 | GDO2
D6 | SO(GDO1)
D5 | SCLK
D7 | SI
GND | GND
VCC | VCC | 3.3 Volt !!!

### Setup instructions

The configuration of the wifi Dongle is stored in the EEPROM memory of the ESP8266. On first initialisation, when no configuration is found, it is initialized with some default values and the Dongle turns on the Admin-Mode.

In Admin-Mode, the blue LED on the ESP submodule is turned on and the Dongle creates an WLAN-Access-Point with the SSID `Shutter-RC-Dongle`, protectet with the WPA-Passwort `12345678`. Now you have 180 seconds (3 minutes) time to connect to the WLAN Accesspoint and visit the configuration webserver on

http://192.168.4.1

Admin-Mode quits after the 180 second timeout or when you restart the Dongle from the configuration webserver.

If you need the Admin-Mode later, just press the "Reset" button on the NodeMCU module two times within 10 seconds. This double-reset will be detected and the Dongle enters Admin-Mode again, showing this with the blue LED turned on.

The running shutter rc wifi Dongle does some debug output on the serial console.
Console Speed is 115200 Bit/s


### Known issues

* after flashing the NodeMCU board, on the first run of the fresh flashed sketch, any kind of restart (through WebUI or after a crash) of the ESP8266 may not properly work, causes the NodeMCU board board to hang. On the serial monitor, you see output like
```
handleFileRead: /favicon.png
 [INFO] - Writing Config to EEPROM

 ets Jan  8 2013,rst cause:2, boot mode:(1,6)


 ets Jan  8 2013,rst cause:4, boot mode:(1,6)

wdt reset
```
* this is a hardware issue, solution: power cycle your NodeMCU after flashing, pressing the reset button is not sufficient

* sometimes opening the "system" page in the WebUI causes a crash (Exception 28) of the ESP8266. The cause is not found yet.

### Contribute

You can contribute to Shutter RC Wifi dongle by
- Truyt to include Oregon sensor in RCSwitch library
- providing Pull Requests (Features, Proof of Concepts, Language files or Fixes)
- testing new released features and report issues
- please try to follow the [Style-Guide](https://www.gnu.org/prep/standards/html_node/Writing-C.html#Writing-C)
  [another Style-Guide](https://google.github.io/styleguide/cppguide.html)


CHANGES.md
## Changelog
**Usage note:** please add new changelog information here at the beginning

##### 2021-01-05
* Add RCSwitch library to receive and send RF 433 signal

#### 2020-12-30
* Remove CC1101.h library only run with ELECHOUSE_CC1101_SRC_DRV

##### 2020-12-20
* Add Somfy shutter suport with ELECHOUSE library

##### 2020-04-12
* merged github pull requests #22 and #23
*  #22 delphin0850 committed on 25 Mar 2019: HTML Fix to save the parameters
*  #23 mvn23 committed on 13 Apr 2019: Fix serial generating and parsing
* update library "pubsubclient" to v2.7

##### 2018-08-09
* now it's possible to use the dongle only via the WebUI, by configure NO MQTT broker 
* a heartbeat function lets blink the NodeMCU/ESP's LED every 4sec

##### 2018-07-07 v0.7
* rework use of web_log_message[] array, now a circular buffer
* fix logging of "incoming MQTT command unknown"

##### 2018-07-01
* merge https://github.com/madmartin/Jarolift_MQTT/pull/5/ (partially)
* adds SNTP time fetch and timestamps in log

##### 2018-06-24 v0.6
* make debug output on serial console configurable
* fix Firefox issue in edit channel names (thanks to sidddy)
* add mqtt "sendconfig" command: "cmd/jarolift/sendconfig". the dongle replies with json encoded config data
* check and confirm running with ESP8266 core 2.4.1
* WebUI: enhance layout of system menu, display version
* system menu: serial number prefix is displayed and configurable in hexadecimal now
* system menu: new field: device counter now configurable

##### 2018-06-09 v0.6-rc2
* add logging of WiFi connect and disconnect events. WiFi reconnect works well when compiled with ESP8266 core 2.3.0

##### 2018-06-05 v0.6-rc2
* send device counter to MQTT server after each command
* add subdirectory /doc for misc documentation, beginning with file describing EEPROM memory layout
* implement new connect logic for MQTT, which makes an effective reconnect
* add "uptime in seconds" timestamp to log

##### 2018-06-01 v0.6-rc1
* versionize config data in EEPROM
* introduce configurable mqtt devicetopic
* fix LWT message handling when devicetopic is reconfigured
* Web-UI:
  * explain MQTT better
  * add example for Home Assistant
  * on shutter page: add channelnumber to each channel
* v0.6-rc1
* program version is logged on startup

##### 2018-05-24
* add LWT to mqtt connect
* mqtt client-id must be unique in the network, so add code to append ESP-ID to default
  client-id to make it unique

##### 2018-05-18
* Admin-Mode changed to Access-Point only
* add double reset detector to re-enter Admin-Mode after initial setUpClass
* fixed lots of typos
* fixed inconsistent commenting style
* fixes wording in config webserver

##### 2018-05-15 v0.5
* changed length of SSID to 32 and WPA passwort to 64. needs reconfiguration of MQTT port number
* insert GPLv3 licensing information into all source files

##### 2017-11-11 v0.4a
* improved senden-routine to prevent wdt to reset the device and get more compatibility (controlling and learing).

##### 2017-11-02
* added a switch (bool regular_learn = true; )to choose which learn method to use.
* Regular method is Up+Down at the same time followed by a stop.

##### 2017-10-30
* fixed an error which prevented to learn a motor to the dongle
* reconfigured wifi as STA.

##### 2017-08-30 v0.4
* Added ESP-Reset via fhem. Added shade command to set the shutter to a predefined position.
* Position is set via the "ssp(channel)" cmd.
* Added Group Reception.

##### 2017-08-19 v0.3
* Added WiFi reconnect routine. (THX to Markus)

##### 2017-07-28 v0.2
* Modified rx/tx switch for not getting stuck (line 1182/1192)

##### 2017-07-20 v0.1
* added reception of jarolift remotes and fhem status update



ELECHOUSE_CC1101_SRC_DRV.h
/*
  ELECHOUSE_CC1101.cpp - CC1101 module library
  Copyright (c) 2010 Michael.
    Author: Michael, <www.elechouse.com>
    Version: November 12, 2010

  This library is designed to use CC1101/CC1100 module on Arduino platform.
  CC1101/CC1100 module is an useful wireless module.Using the functions of the 
  library, you can easily send and receive data by the CC1101/CC1100 module. 
  Just have fun!
  For the details, please refer to the datasheet of CC1100/CC1101.
----------------------------------------------------------------------------------------------------------------
cc1101 Driver for RC Switch. Mod by Little Satan. With permission to modify and publish Wilson Shen (ELECHOUSE).
----------------------------------------------------------------------------------------------------------------
*/
#ifndef ELECHOUSE_CC1101_SRC_DRV_h
#define ELECHOUSE_CC1101_SRC_DRV_h

#include <Arduino.h>

//***************************************CC1101 define**************************************************//
// CC1101 CONFIG REGSITER
#define CC1101_IOCFG2       0x00        // GDO2 output pin configuration
#define CC1101_IOCFG1       0x01        // GDO1 output pin configuration
#define CC1101_IOCFG0       0x02        // GDO0 output pin configuration
#define CC1101_FIFOTHR      0x03        // RX FIFO and TX FIFO thresholds
#define CC1101_SYNC1        0x04        // Sync word, high INT8U
#define CC1101_SYNC0        0x05        // Sync word, low INT8U
#define CC1101_PKTLEN       0x06        // Packet length
#define CC1101_PKTCTRL1     0x07        // Packet automation control
#define CC1101_PKTCTRL0     0x08        // Packet automation control
#define CC1101_ADDR         0x09        // Device address
#define CC1101_CHANNR       0x0A        // Channel number
#define CC1101_FSCTRL1      0x0B        // Frequency synthesizer control
#define CC1101_FSCTRL0      0x0C        // Frequency synthesizer control
#define CC1101_FREQ2        0x0D        // Frequency control word, high INT8U
#define CC1101_FREQ1        0x0E        // Frequency control word, middle INT8U
#define CC1101_FREQ0        0x0F        // Frequency control word, low INT8U
#define CC1101_MDMCFG4      0x10        // Modem configuration
#define CC1101_MDMCFG3      0x11        // Modem configuration
#define CC1101_MDMCFG2      0x12        // Modem configuration
#define CC1101_MDMCFG1      0x13        // Modem configuration
#define CC1101_MDMCFG0      0x14        // Modem configuration
#define CC1101_DEVIATN      0x15        // Modem deviation setting
#define CC1101_MCSM2        0x16        // Main Radio Control State Machine configuration
#define CC1101_MCSM1        0x17        // Main Radio Control State Machine configuration
#define CC1101_MCSM0        0x18        // Main Radio Control State Machine configuration
#define CC1101_FOCCFG       0x19        // Frequency Offset Compensation configuration
#define CC1101_BSCFG        0x1A        // Bit Synchronization configuration
#define CC1101_AGCCTRL2     0x1B        // AGC control
#define CC1101_AGCCTRL1     0x1C        // AGC control
#define CC1101_AGCCTRL0     0x1D        // AGC control
#define CC1101_WOREVT1      0x1E        // High INT8U Event 0 timeout
#define CC1101_WOREVT0      0x1F        // Low INT8U Event 0 timeout
#define CC1101_WORCTRL      0x20        // Wake On Radio control
#define CC1101_FREND1       0x21        // Front end RX configuration
#define CC1101_FREND0       0x22        // Front end TX configuration
#define CC1101_FSCAL3       0x23        // Frequency synthesizer calibration
#define CC1101_FSCAL2       0x24        // Frequency synthesizer calibration
#define CC1101_FSCAL1       0x25        // Frequency synthesizer calibration
#define CC1101_FSCAL0       0x26        // Frequency synthesizer calibration
#define CC1101_RCCTRL1      0x27        // RC oscillator configuration
#define CC1101_RCCTRL0      0x28        // RC oscillator configuration
#define CC1101_FSTEST       0x29        // Frequency synthesizer calibration control
#define CC1101_PTEST        0x2A        // Production test
#define CC1101_AGCTEST      0x2B        // AGC test
#define CC1101_TEST2        0x2C        // Various test settings
#define CC1101_TEST1        0x2D        // Various test settings
#define CC1101_TEST0        0x2E        // Various test settings

//CC1101 Strobe commands
#define CC1101_SRES         0x30        // Reset chip.
#define CC1101_SFSTXON      0x31        // Enable and calibrate frequency synthesizer (if MCSM0.FS_AUTOCAL=1).
                                        // If in RX/TX: Go to a wait state where only the synthesizer is
                                        // running (for quick RX / TX turnaround).
#define CC1101_SXOFF        0x32        // Turn off crystal oscillator.
#define CC1101_SCAL         0x33        // Calibrate frequency synthesizer and turn it off
                                        // (enables quick start).
#define CC1101_SRX          0x34        // Enable RX. Perform calibration first if coming from IDLE and
                                        // MCSM0.FS_AUTOCAL=1.
#define CC1101_STX          0x35        // In IDLE state: Enable TX. Perform calibration first if
                                        // MCSM0.FS_AUTOCAL=1. If in RX state and CCA is enabled:
                                        // Only go to TX if channel is clear.
#define CC1101_SIDLE        0x36        // Exit RX / TX, turn off frequency synthesizer and exit
                                        // Wake-On-Radio mode if applicable.
#define CC1101_SAFC         0x37        // Perform AFC adjustment of the frequency synthesizer
#define CC1101_SWOR         0x38        // Start automatic RX polling sequence (Wake-on-Radio)
#define CC1101_SPWD         0x39        // Enter power down mode when CSn goes high.
#define CC1101_SFRX         0x3A        // Flush the RX FIFO buffer.
#define CC1101_SFTX         0x3B        // Flush the TX FIFO buffer.
#define CC1101_SWORRST      0x3C        // Reset real time clock.
#define CC1101_SNOP         0x3D        // No operation. May be used to pad strobe commands to two
                                        // INT8Us for simpler software.
//CC1101 STATUS REGSITER
#define CC1101_PARTNUM      0x30
#define CC1101_VERSION      0x31
#define CC1101_FREQEST      0x32
#define CC1101_LQI          0x33
#define CC1101_RSSI         0x34
#define CC1101_MARCSTATE    0x35
#define CC1101_WORTIME1     0x36
#define CC1101_WORTIME0     0x37
#define CC1101_PKTSTATUS    0x38
#define CC1101_VCO_VC_DAC   0x39
#define CC1101_TXBYTES      0x3A
#define CC1101_RXBYTES      0x3B

//CC1101 PATABLE,TXFIFO,RXFIFO
#define CC1101_PATABLE      0x3E
#define CC1101_TXFIFO       0x3F
#define CC1101_RXFIFO       0x3F

#define CC1101_CFG_SIZE 0x2F //47  
 
  const uint8_t CC1101_JAROLIFT_RTS_CFG[CC1101_CFG_SIZE] = {
  // Jarolif IDX NAME COMMENT
  0x0D,// 00 IOCFG2 GDO2 as serial output
  0x2E,// 01 IOCFG1 Tri-State
  0x2D,// 02 IOCFG0 GDO0 for input
  0x47,// 03 FIFOTHR 
  0xD3,// 04 SYNC1 
  0x91,// 05 SYNC0 
  0x3D,// 06 PKTLEN 
  0x04,// 07 PKTCTRL1 
  0x32,// 08 PKTCTRL0 
  0xFF,// 09 ADDR 
  0x00,// 0A CHANNR 
  0x06,// 0B FSCTRL1 152kHz IF Frquency
  0x00,// 0C FSCTRL0 
  0x10,// 0D FREQ2 433.42 (10,AB,85: Somfy RTS Frequency)
  0xB0,// 0E FREQ1 
  0x71,// 0F FREQ0 
  0x69,// 10 MDMCFG4 bWidth 325kHz
  0xE4,// 11 MDMCFG3 Drate: 828 ((256+11)*2^5)*26000000/2^28
  0x30,// 12 MDMCFG2 Modulation: ASK
  0x22,// 13 MDMCFG1 
  0xF8,// 14 MDMCFG0 ChannelSpace: 350kHz
  0x00,// 15 DEVIATN 
  0x07,// 16 MCSM2 
  0x30,// 17 MCSM1 
  0x18,// 18 MCSM0 Calibration: RX/TX->IDLE
  0x14,// 19 FOCCFG 
  0x6C,// 1A BSCFG 
  0x04,// 1B AGCCTRL2 42 dB instead of 33dB
  0x00,// 1C AGCCTRL1 
  0x92,// 1D AGCCTRL0 4dB decision boundery
  0x87,// 1E WOREVT1 
  0x6B,// 1F WOREVT0 
  0xF8,// 20 WORCTRL 
  0xB6,// 21 FREND1 
  0x11,// 22 FREND0 0x11 without Pa ramping
  0xE9,// 23 FSCAL3 
  0x2A,// 24 FSCAL2 
  0x00,// 25 FSCAL1 
  0x1F,// 26 FSCAL0 
  0x41,// 27 RCCTRL1 
  0x00,// 28 RCCTRL0 
  0x59,// 29 FSTEST
  0x7F,// 2A PTEST
  0x3F,// 2B AGCTEST
  0x81,// 2C TEST2
  0x35,// 2D TEST1
  0x09// 2E TEST0
  };


//************************************* class **************************************************//
class ELECHOUSE_CC1101
{
private:
  void SpiStart(void);
  void SpiEnd(void);
  void GDO_Set (void);
  void Reset (void);
  void setSpi(void);
  void RegConfigSettings(void);
  void Calibrate(void);
  void Split_PKTCTRL0(void);
  void Split_PKTCTRL1(void);
  void Split_MDMCFG1(void);
  void Split_MDMCFG2(void);
  void Split_MDMCFG4(void);
public:
  void Init(void);
  byte SpiReadStatus(byte addr);
  void setSpiPin(byte sck, byte miso, byte mosi, byte ss);
  void setGDO(byte gdo0, byte gdo2);
  void setCCMode(bool s);
  void setModulation(byte m);
  void setPA(int p);
  void setMHZ(float mhz);
  void setChannel(byte chnl);
  void setChsp(float f);
  void setRxBW(float f);
  void setDRate(float d);
  void setDeviation(float d);
  void SetTx(void);
  void SetRx(void);
  void SetTx(float mhz);
  void SetRx(float mhz);
   int getRssi(void);
  byte getLqi(void);
  void setSres(void);
  void SendData(byte *txBuffer, byte size);
  void SendData(char *txchar);
  void SendData(byte *txBuffer, byte size, int t);
  void SendData(char *txchar, int t);
  byte CheckReceiveFlag(void);
  byte ReceiveData(byte *rxBuffer);
  bool CheckCRC(void);
  void SpiStrobe(byte strobe);
  void SpiWriteReg(byte addr, byte value);
  void SpiWriteBurstReg(byte addr, byte *buffer, byte num);
  byte SpiReadReg(byte addr);
  void SpiReadBurstReg(byte addr, byte *buffer, byte num);
  void setClb(byte b, byte s, byte e);
  void setSyncWord(byte sh, byte sl);
  void setAddr(byte v);
  void setWhiteData(bool v);
  void setPktFormat(byte v);
  void setCrc(bool v);
  void setLengthConfig(byte v);
  void setPacketLength(byte v);
  void setDcFilterOff(bool v);
  void setManchester(bool v);
  void setSyncMode(byte v);
  void setFEC(bool v);
  void setPQT(byte v);
  void setCRC_AF(bool v);
  void setAppendStatus(bool v);
  void setAdrChk(byte v);
  bool CheckRxFifo(int t);
};

extern ELECHOUSE_CC1101 ELECHOUSE_cc1101;

#endif
